integration-tests/tests/src/tests/queries.ts

```diff
@@ -15,11 +15,865 @@
 // limitations under the License.
 //
 ////////////////////////////////////////////////////////////////////////////
-import { expect } from "chai";
 import Realm from "realm";
+import { expect } from "chai";
+const { Decimal128, ObjectId, UUID } = Realm.BSON;
+
+interface SchemaWithPropertyArray extends Omit<Realm.ObjectSchema, "properties"> {
+  properties: { name: string; type: string; objectType?: string; optional?: boolean }[];
+}
+
+type AssertLengthTest = [
+  TestType: string,
+  ObjectType: string,
+  ExpectedLength: number,
+  Query: string,
+  ...QueryArgs: Array<any>
+];
+
+type AssertExceptionTest = [
+  TestType: string,
+  ObjectType: string,
+  ExpectedException: string,
+  Query: string,
+  ...QueryArgs: Array<any>
+];
+
+type AssertResultValuesTest = [
+  TestType: string,
+  ObjectType: string,
+  ExpectedValues: any[],
+  Query: string,
+  ...QueryArgs: Array<any>
+];
+
+interface TestCase {
+  schema: SchemaWithPropertyArray[];
+  objects: {
+    type: string;
+    value: any[];
+  }[];
+  tests: (AssertLengthTest | AssertExceptionTest | AssertResultValuesTest)[];
+}
+
+const SharedTestSuiteCases: Record<string, TestCase> = {
+  dateTests: {
+    schema: [{ name: "DateObject", properties: [{ name: "dateCol", type: "date" }] }],
+    objects: [
+      { type: "DateObject", value: [0] },
+      { type: "DateObject", value: [1] },
+      { type: "DateObject", value: [2] },
+    ],
+    tests: [
+      ["AssertLength", "DateObject", 2, "dateCol < $0", [2, "dateCol"]],
```

Comment:
We definitely need to refactor this.  Anyone starting to read this code from top to bottom will become immediately clueless as to what any of these tests means.  What may be a good alternative before merging is to add some comments about what these array values actually do.  This could help with future refactoring work.